kafka-event-listener/src/main/java/org/hypertrace/core/kafka/event/listener/KafkaEventListener.java

```diff
@@ -1,21 +1,38 @@
 package org.hypertrace.core.kafka.event.listener;
 
+import com.typesafe.config.Config;
 import java.time.Duration;
+import java.util.ArrayList;
 import java.util.List;
+import org.apache.kafka.common.serialization.Deserializer;
 
 public abstract class KafkaEventListener<K, V> {
-  private final List<KafkaEventConsumer<K, V, Object, Object>> eventConsumers;
+  private final List<KafkaEventListenerConsumer<K, V, Object, Object>> eventListenerConsumers =
+      new ArrayList<>();
 
-  public KafkaEventListener(List<KafkaEventConsumer<K, V, Object, Object>> eventConsumers) {
-    this.eventConsumers = eventConsumers;
-    eventConsumers.forEach(Thread::start);
+  public <D, C> void registerConsumer(
+      String consumerName,
+      Config kafkaConfig,
+      Deserializer<D> keyDeserializer,
+      Deserializer<C> valueDeserializer) {
+    registerConsumer(
+        new KafkaEventListenerConsumer<>(
+            consumerName, kafkaConfig, keyDeserializer, valueDeserializer, this));
   }
 
+  /** only for test usage */
```

Comment:
Thanks for the detailed explanation

> Basically your consumer could be exposed as an interface that has to implement 2 things only - start and stop

Yes it could be exposed as such an interface with start, stop and Kafka based is one such implementation. Any subscription type client is an ideal implementation of that interface (say streaming grpc method).

The logical question now is, is this repo the right place to host this listener and KafkaConsumer impl? Once we create an interface for abstraction I don't see the need to prefix the Listener or the Consumer interface with Kafka. Kafka just becomes an impl detail for one such implementation of the interface

Reply:
Yeah, it's an interesting question. To take it a step further, you've also posed bidirectional scenarios (reusing the same consumer across multiple listeners; giving multiple consumers to the same listener). I'd encourage you to go review the change (if you haven't already) ignoring the kafka impl - there's not too much left. It's basically just responsible for linking the producer and consumer. And extremely confusingly, the listener here is the consumer (it "consumes" via it's `actOnEvent`) and the consumer is actually the producer (while the abstraction is neither - it just starts and stops - in the kafka impl you produce to the `actOnEvent` method handed to it - the consumption from kafka is just part of the impl). So it seems like we're creating a mini local message bus - so when we frame it like that, does it still make sense? A registry may be a better fit, but there's also many third party libs that already do this very robustly I'm sure.

Happy to hop on a call tomorrow with you if you want to dig in.